### PR TITLE
feat: add proper handling of query/path/body parameters for rest transport

### LINKED ISSUE #267: Re-add HTTP transport and make it the fallback
Some users may not want to use gRPC. Http is a sufficient fallback for certain APIs, so this transport must be explicitly supported.
This support used to exist but was removed due to testing limitations.
See https://github.com/googleapis/gapic-generator-python/pull/114 for additional details.

### PR DESCRIPTION
**Pull request by yon-mg**
Implementation and tests assume basic case of grpc transcoding for handling of all query/path parameters and body field.

These changes are dependent upon a PR in googleapis/proto-plus-python repo:
https://github.com/googleapis/proto-plus-python/pull/164

### CODE REVIEW

File: gapic/schema/wrappers.py
Comment (software-dov) [inline gapic/schema/wrappers.py:784]: We can move the variable setups around a little to make it a little easier to read.
```python
params = set(self.path_params)
body = self.http_opt.get('body')
if body:
    params.add(body)

return set(self.input.fields) ^ params
```
Operations on dicts by default just use the keyspace, e.g. `assert 6 in {6: "six"}` and `assert set({6: "six"}) == {6}`.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:169]: Nit: prefer format strings to concatenating using +. It's easier to eyeball parse for longer or more convoluted strings.
```python
url += "{q}{name}={value}".format(q=q, name=param_name, value=param_value.replace(' ', '+'))
```

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:173]: No need to use `keys` here, `if 'body' in method.http_opt` is more idiomatic.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:166]: Very minor premature optimization nit: since we're immediately iterating over and then discarding this dictionary, we can turn it into a generator instead and prevent looping over the same data multiple times.
```python
potential_params = ((k, v) for k, v in potentialParams.items() if v) # The parentheses make this a generator expression.
for i, (param_name, param_value) in enumerate(potentialParams):
```

[This](https://wiki.python.org/moin/Generators) is a good rundown on generators and generator expressions. Dave Beazley also has a really fun youtube talk on generators.

File: gapic/utils/case.py
Comment (software-dov) [inline gapic/utils/case.py]: Nit: no need to make a list as the argument to join. We could replace the square brackets with parentheses to make it a generator expression, but there's a minor syntactic optimization where if a generator expression is the sole function argument you can remove the parens.
```python
join(x.capitalize() for x in items[1:])
```

File: gapic/generator/generator.py
Comment (software-dov) [inline gapic/generator/generator.py:293]: Just a reminder to clean this up.

**Review by software-dov — COMMENTED** [review]
Looks good so far, just took an early readthrough.

File: gapic/utils/case.py
Comment (vam-google) [inline gapic/utils/case.py:49]: Are there any built-in/standard python functions which do the same? Please prefer using standard ones to custom, if there are any.

File: gapic/schema/wrappers.py
Comment (vam-google) [inline gapic/schema/wrappers.py:774]: Please add a comment that this handles grpc encoding in its simples case

**Review by vam-google — COMMENTED** [review]
Overall looks good, please try to get LGTM when Dov comes back and push this

File: gapic/utils/case.py
Comment (software-dov) [reply gapic/utils/case.py:49]: I'm fairly certain there is no standard library function for to camel-case.

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:782]: !! Q: can you know all the query parameters ahead of time? If a user sets a field nested inside a message that is not listed in the path template nor in the body part of the annotation, it should be reported as a query param, but presumably in dotted notation (see my other comment about dotted notation being used in practice for the body even though disallowed in theory).

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:139]: nit: maybe have the generated code call this variable `body` rather than `data` so the code is more self-explanatory.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:169]: !! Does this ensure values are URL-encoded? If not, at least add a TODO.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:163]: Sounds like this should work for nested fields (dot notation). Correct?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:166]: It would probably be clearer to s/`potentialParams`/`queryParams`/

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:161]: It would probably be clearer to s/`potentialParams`/`queryParams`/g

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:166]: This is fine, but another approach is to `potentialParams=["{k}={v}".format(k=k, v=v) for k, v in ...]` here and then `url += "?{}".format("&".join(potentialParams))` below

Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:166]: !! re: `if v`
Is the Python truthiness of `v` what we want here? I don't know, but I wonder given that we want to transmit only explicitly-set values, even if they match the defaults, which are falsy.

For any data type in the Python data model:
- if the unset value is null, then we can test for null to determine whether to send it. Simply checking for falsiness would not make this distinction.
- if the unset value is non-null, is there some other way to distinguish between values that were explicitly set and those that weren't?

We should talk about this, and at the very least insert a TODO for now.

File: gapic/utils/case.py
Comment (vchudnov-g) [inline gapic/utils/case.py:51]: nit: In this and the previous, pre-existing function, we're not touching spaces, right? Might be worth mentioning that.

File: gapic/utils/case.py
Comment (vchudnov-g) [inline gapic/utils/case.py:62]: I'm surprised by the hyphen `-`. It almost seems like that should be handled by `to_snake_case`, though that's outside the scope of this PR. Does it come up in practice, that an input to this function has hyphens/kebab-case?

Opinion, @software-dov ?

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:781]: Just curious: why can't  `query_params`  and `path_params` return the same type, rather than a `Set` and a `Sequence`?

**Review by vchudnov-g — CHANGES_REQUESTED** [review]
I flagged my more important comments with `!!`

It looks like this generator does not have golden files or baselines, correct? Too bad; it would be good to see a sample of the generated GAPIC showing off these features.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:139]: It does not handle it but I have added a todo in the appropriate place in `wrappers.Method`. Pretty much anything dealing with full handling of grpc transcoding (or special cases outside grpc transcoding) is not yet taken care of in this PR.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:166]: Short answer is yes and your second bullet point applies. Afaik from debugging it, there isn't actually a way to tell if a value was unset or set to default value since when the `<InputType>Request` object is made and passed to this generated method as a parameter, its values seem set to default ones. I also believe this is intended. For example, if I made a request but set a query param value to an empty string, simply checking if a value was set would allow that empty string to be included as part of the path_url and may result in a `'Bad Request'` response.

File: gapic/schema/wrappers.py
Comment (yon-mg) [reply gapic/schema/wrappers.py]: I agree you're right and I'll change it but there shouldn't be a difference between difference and symmetric difference here since all `body + path_params` should implicitly be included in all_params. I suppose that's why it hasn't failed for me but I'm not sure why I didn't just use `-` in the first place.

Comment (yon-mg) [reply gapic/schema/wrappers.py]: Not really here no but I have a todo for that as well as including in the pr description.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:169]: Good point! Will do that

File: gapic/utils/case.py
Comment (yon-mg) [reply gapic/utils/case.py:62]: It actually might handle it: I didn't check. As far as use, it doesn't get used but I wouldn't want a future user to attempt o try it and have it fail.

File: gapic/schema/wrappers.py
Comment (yon-mg) [reply gapic/schema/wrappers.py:781]: I suppose they could. Just thought it's more appropriate since ordering seems important for `path_params` but not for `query_params`.

File: gapic/utils/case.py
Comment (software-dov) [reply gapic/utils/case.py:62]: My vote: write a test for `to_snake_case`'s behavior, add the functionality if the test fails, just split on '_'.

File: gapic/schema/wrappers.py
Comment (software-dov) [reply gapic/schema/wrappers.py]: I have no strong feeling for whether we want difference or XOR; Victor and the spec should take priority there. +1 for adding a test that explicitly checks for the one that we actually want and NOT the other one.

Side note: IMHO the set operators are not very clear, and especially if doing more complex set operations it's more reader friendly to use the named methods.

File: gapic/schema/wrappers.py
Comment (software-dov) [inline gapic/schema/wrappers.py:778]: We can move more of the logic into the regex by using capturing.
```python
param_re = r"\{(\w+)\}"
return re.findall(param_re, self.http_opt['url']) if self.http_opt else []
```

File: gapic/utils/case.py
Comment (vchudnov-g) [reply gapic/utils/case.py:62]: If `to_snake_case` does not handle it, could adding the functionality break existing clients? is there an easy way to check? (This is why I was suggesting "outside the scope of this PR")

Comment (yon-mg) [reply gapic/utils/case.py:62]: I doubt it would break anything since it's additive but this is why I didn't change `to_snake_case`'s behavior

File: gapic/schema/wrappers.py
Comment (software-dov) [reply gapic/schema/wrappers.py]: Replying to Victor's ancestral comment: the generated unit tests and 100% coverage thereof is intended to fill a similar goal as golden tests without making people tear their hair out from frustration. Occasionally it is helpful to diff surfaces, though; I like to either stash changes or do sub-branches, invoke the generator on showcase to save the resulting GAPIC, then ediff the relevant files in the two outputs. It can be a little fiddly, but it's *_far_* easier to do as a tactical move than to have golden tests.

Comment (vchudnov-g) [reply gapic/schema/wrappers.py]: For reviews, it's _very_ helpful to be able to diff the generated surfaces. That's the main thing I miss about golden files. The frustration, I can do without. ;-)

Comment (vchudnov-g) [reply gapic/schema/wrappers.py]: For this particular comment, as Yon and I discussed yesterday: `body + list(self.path_params)` should be a subset of `all_params`, so difference and xor should behave the same. However, difference expresses the intent better and it makes it more likely we can catch or prevent errors if this assumption is violated due to a bug (now or in the future).

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:139]: OK. As discussed, we should err on the side of duplicate TODOs rather than too few. If we address one, we're likely to see the others and address/delete them.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:144]: As we clarified yesterday: params in the path may be (but don't have to be) repeated in the body if we're using `"*"`, though we should never have a `body: "foo"` if `foo` is a path param. And query params are whatever's left over from what's not included in the path and body. So we just need to ensure no body params wind up as query params.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:166]: OK, this ties back to the default values design I still have open.....

On empty strings, maybe  one of the [ideas from the web](https://www.google.com/search?q=empty+strings+as+query+params&oq=empty+strings+as+query+params&aqs=chrome..69i57j69i64l3.6119j1j7&sourceid=chrome&ie=UTF-8) will work with OP APIs. But we probably don't need to spend much time investigating until we decide on a the approach to default values. It certainly shouldn't block this PR.

But yes, let's add a TODO that likely something will need to be done here once we decide what we're doing for default values.

File: gapic/utils/case.py
Comment (vchudnov-g) [reply gapic/utils/case.py:62]: But the generator has already produced libraries that were published, so if one of those had a hyphen somewhere, it made it through `to_snake_case` and changing that implementation could lead to an incompatible surface.

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [reply gapic/schema/wrappers.py:781]: As discussed. That seems reasonable, but a complication is that repeated fields map to repeated query params. In that case, a sequence may be best. (If order doesn't matter, a multiset might be enough, but we can't assume that order doesn't matter for repeated fields)

File: gapic/schema/wrappers.py
Comment (yon-mg) [reply gapic/schema/wrappers.py:772]: Well it doesn't do this yet so I'll update the comment when it does.

File: gapic/schema/wrappers.py
Comment (yon-mg) [reply gapic/schema/wrappers.py:782]: This is an oversight yes. Due to these nested fields I don't think it makes sense to identify all query params ahead of time, though it is possible! I am moving this logic out of the generator entirely and have left appropriate todos.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:169]: So the short answer is yes but I will add tests to confirm that.

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [reply gapic/schema/wrappers.py:772]: WDYM? You're returning the `{foo}` which is a field path, right?

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [reply gapic/schema/wrappers.py:778]: good idea

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [reply gapic/schema/wrappers.py:782]: OK. I see you added TODOs to come back to this later. sgtm.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:179]: OK, great. If it's not trivial to URL encode, you can leave a TODO and do it in a follow-up PR, which is also fine. That would let us start to use/test what you have here for well formed data where URL encoding isn't necessary.

Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:179]: (BTW, do you know why the `&` shows up in red on GitHub?)

**Review by vchudnov-g — COMMENTED** [review]
Looking good. Two of my previous `!!` comments are still open. The truthiness one we should mark with a TODO, since we can't fully resolve it right now. The URL-encoding one, if it's not trivial to address now, we should mark with a TODO and address in a follow-up PR.

File: gapic/schema/wrappers.py
Comment (yon-mg) [reply gapic/schema/wrappers.py:781]: You are right set is not appropriate even if order doesn't mater but I should mention though that once query param logic is moved out, this won't matter anymore.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:179]: Not sure. Seems to be a bug w/ ampersand messing with syntax highlighting? Certainly no real error AFAIK.

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [reply gapic/schema/wrappers.py:781]: True!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:182]: Thanks for the TODO!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:176]: Thanks for the TODO on truthiness!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:180]: nit: Do consider the string.Join suggestion from earlier (but it's certainly not a blocker)

**Review by vchudnov-g — APPROVED** [review]
Looks good! Thanks for doing this!

For clarity when looking back at this commit, maybe you can expand in the merge commit description what the "basic case" is: path params in simple form, no body, no query param.

**Review by software-dov — APPROVED** [review]

File: gapic/schema/wrappers.py
Comment (software-dov) [reply gapic/schema/wrappers.py]: Excellent. Glad to hear that the issue has been resolved.

**Pull request merged**
